test/test_source.cpp

```diff
@@ -0,0 +1,187 @@
+                    "#include \"cuda_runtime.h\"\n"
+                    "#include \"test_compare.hpp\"\n"
+                    "extern \"C\"\n"
+                    "__global__ void cuda_add_with_header(int *x, int *y, int *out, int"
+                    " datasize) {\n"
+                    " compare check = OUT_COMPARE_WRONG; \n"
+                    " int i = threadIdx.x;\n"
+                    " out[i] = x[i] + y[i] + 5;\n"
+                    " if(out[i]%5==0) check = CORRECT;\n"
+                    "}", headersNew1};
+
+                dim3 gridNew1(1);
+                dim3 blockNew1(5);
+                
+                sourceNew1.program("cuda_add_with_header").compile().configure(gridNew1, blockNew1).launch(args);
+                
+                //B3. Comparing the results
+                for (int i=0;i<5;i++) REQUIRE(hOut[i] == hostCompareOutput_1[i]);
+            }    
+        }
+
+        //C1. Checking for consistencies for running various functions in a kernel-code
+        WHEN("More one functions in a kernel-code are used."){
+            args.clear();
+            args.emplace_back(KernelArg{hX, bufferSize});
+            args.emplace_back(KernelArg{hY, bufferSize});
+            args.emplace_back(KernelArg{hOut_Multiply1, bufferSize, true});
+            args.emplace_back(KernelArg(&datasize));
+            
+            Headers headersNew2;
+            headersNew2.insert(Header{"cuda_runtime.h"});
+
+            Source sourceNew2{
+                "#include \"cuda_runtime.h\"\n"
+                "extern \"C\"\n"
+                "__global__ void cuda_add_normal(int *x, int *y, int *out, int "
+                "datasize) {\n"
+                " int i = threadIdx.x;\n"
+                " out[i] = x[i] + y[i];\n"
+                "}\n"
+                "extern \"C\"\n"
+                "__global__ void cuda_add_five(int *x, int *y, int *out, int"
+                " datasize) {\n"
+                " int i = threadIdx.x;\n"
+                " out[i] = x[i] + y[i] + 5;\n"
+                "}\n", headersNew2};
+
+            /*C2. Compilation of Kernels through the creation of a 
+            program from kernel - source*/
+            THEN("The results are consistent with the given controlled output."){
+                dim3 gridNew2(1);
+                dim3 blockNew2(5);
+                
+                sourceNew2.program("cuda_add_five").compile().configure(gridNew2, blockNew2).launch(args);
+                
+                //C3. Compiling another function in a kernel-code.
+                args.clear();
+                args.emplace_back(KernelArg{hX, bufferSize});
+                args.emplace_back(KernelArg{hY, bufferSize});
+                args.emplace_back(KernelArg{hOut_Multiply2, bufferSize, true});
+                args.emplace_back(KernelArg(&datasize));   
+               
+                sourceNew2.program("cuda_add_normal").compile().configure(gridNew2, blockNew2).launch(args);
```

Comment:
this is nice, I just wish you also tested if an exception is thrown if the function name is wrong, e.g. if you did:
```c++
sourceNew2.program("this_function_does_not_exist").compile()
```

Reply:
I'm not really sure if this test `Source::program()` or `Program::compile()` more.
If you want you could pretty much remove everything from this file and just check if you can call `.program()` and put all the interesting checks to #115